repo.go

```diff
@@ -694,28 +975,88 @@ func (r *Repo) AddTargetsWithDigest(digest string, digestAlg string, length int6
 
 	t.Targets[path] = data.TargetFileMeta{FileMeta: meta}
 
-	return r.writeTargetWithExpires(t, expires)
+	return r.writeTopLevelTargetWithExpires(t, expires)
 }
 
 func (r *Repo) AddTargetWithExpires(path string, custom json.RawMessage, expires time.Time) error {
 	return r.AddTargetsWithExpires([]string{path}, custom, expires)
 }
 
 func (r *Repo) AddTargetsWithExpires(paths []string, custom json.RawMessage, expires time.Time) error {
+	return r.AddTargetsWithExpiresToPreferredRole(paths, custom, expires, "")
+}
+
+func (r *Repo) AddTargetWithExpiresToPreferredRole(path string, custom json.RawMessage, expires time.Time, preferredRole string) error {
+	return r.AddTargetsWithExpiresToPreferredRole([]string{path}, custom, expires, preferredRole)
+}
+
+type targetsMetaWithSigners struct {
+	meta    *data.Targets
+	signers []keys.Signer
+}
+
+// AddTargetsWithExpiresToPreferredRole signs the staged targets at `paths`. If
+// preferredRole is not nil, the target is added to the given role's manifest
+// if delegations allow it. If delegations do not allow the preferredRole to
+// sign the given path, an error is returned.
+func (r *Repo) AddTargetsWithExpiresToPreferredRole(paths []string, custom json.RawMessage, expires time.Time, preferredRole string) error {
 	if !validExpires(expires) {
 		return ErrInvalidExpires{expires}
 	}
 
-	t, err := r.topLevelTargets()
-	if err != nil {
-		return err
-	}
 	normalizedPaths := make([]string, len(paths))
 	for i, path := range paths {
 		normalizedPaths[i] = util.NormalizeTarget(path)
 	}
+
+	// As we iterate through staged files, we accumulate changes to their
+	// corresponding targets manifests.
+	targetsMetaToWrite := map[string]*targetsMetaWithSigners{}
+
 	if err := r.local.WalkStagedTargets(normalizedPaths, func(path string, target io.Reader) (err error) {
-		meta, err := util.GenerateTargetFileMeta(target, r.hashAlgorithms...)
+		targetsMeta, delegation, err := r.targetDelegationForPath(path, preferredRole)
+		if err != nil {
+			return err
+		}
+		// This is the targets role that needs to sign the target file.
+		targetsRoleName := delegation.Delegatee.Name
+
+		signers, err := r.getSignersInDB(targetsRoleName, delegation.DB)
+		if err != nil {
+			return err
+		}
+
+		// The delegations determine the keys that need to be used to sign each
+		// target file. If a particular target role is reachable in the delegation
+		// graph through multiple paths, it may be the case that the targets
+		// manifest needs signatures from multiple sets of keys.  Therefore, we
+		// merge the keys specified by the current delegation with keys specified
+		// bypreviously seen delegations.
+		if prevMetaWithSigners, ok := targetsMetaToWrite[targetsRoleName]; ok {
```

Comment:
I'm not sure this implementation is correct. I think in order to produce metadata that's verifiable by any valid incoming delegation key set, the metadata needs to be signed by the union of the set of all keys in all the delegations that lead to this role. Currently, I think this is only signing using keys for the first traversed delegation that leads to the `preferredRole` (or if `preferredRole` is not specified, the first delegation that can sign for the given path).

Reply:
This is not applicable if the delegation graph is a tree, but the more general case where a role can have multiple valid delegation key sets adds significant complexity like this.